// @lwrjs/tools/build/es/plugins/json-with.js
// This plugin is a temporary solution to handle JSON imports with the 'with' syntax
// until esbuild adds native support for it. Once esbuild supports the 'with' syntax
// for JSON imports (https://github.com/evanw/esbuild/issues/3384), this plugin can be removed.
// The plugin allows us to use:
// import data from './data.json' with { type: 'json' };
// instead of the older assert syntax:
// import data from './data.json' assert { type: 'json' };
import fs from 'fs-extra';
import path from 'path';
export default function jsonWithPlugin() {
    return {
        name: 'json-with-loader',
        setup(build) {
            build.onLoad({ filter: /.*/ }, async (args) => {
                if (args.with?.type === 'json') {
                    try {
                        const contents = await fs.readFile(args.path, 'utf8');
                        const jsonData = JSON.parse(contents);
                        return {
                            contents: `module.exports = ${JSON.stringify(jsonData)}`,
                            loader: 'js',
                            resolveDir: path.dirname(args.path),
                        };
                    }
                    catch (error) {
                        return {
                            errors: [
                                {
                                    text: `Failed to load JSON file with { type: 'json' }: ${error instanceof Error ? error.message : String(error)}`,
                                    location: { file: args.path },
                                },
                            ],
                        };
                    }
                }
            });
        },
    };
}
