// @lwrjs/core/build/cjs/middleware/request-processor-middleware.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/core/src/middleware/request-processor-middleware.ts
__markAsModule(exports);
__export(exports, {
  requestProcessorMiddleware: () => requestProcessorMiddleware
});
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_config = __toModule(require("@lwrjs/config"));
function requestProcessorMiddleware(app, context) {
  const {basePath} = context.runtimeEnvironment;
  app.use(async (req, res, next) => {
    if (!req.headers) {
      req.basePath = basePath;
      await next();
      return;
    }
    logRequestHeaders(req.headers);
    const requestClass = req.headers[import_shared_utils.MRT_REQUEST_CLASS];
    const requestDepth = (0, import_shared_utils.parseRequestDepth)(req.headers, req.query);
    if ((0, import_shared_utils.isLambdaEnv)()) {
      const forwarded = req.headers["forwarded"];
      const forwardedProto = req.headers["x-forwarded-proto"];
      const host = req.headers["host"];
      const trueClientIP = req.headers[import_shared_utils.TRUE_CLIENT_IP];
      const correlationID = req.headers[import_shared_utils.CORRELATION_ID];
      const protocol = req.protocol;
      const cookieLength = req.headers["cookie"]?.length || 0;
      const cookieMsg = buildCookieMessage(req, cookieLength);
      import_diagnostics.logger.info({
        label: "request-processor-middleware",
        message: `Original Url: ${req.originalUrl}, Forwarded: ${forwarded}, X-Forwarded-Proto: ${forwardedProto}, Host: ${host}, Protocol: ${protocol}, ${cookieMsg}, ${import_shared_utils.MRT_REQUEST_CLASS}: ${requestClass}, ${import_shared_utils.TRUE_CLIENT_IP}: ${trueClientIP}, ${import_shared_utils.CORRELATION_ID}: ${correlationID}, ${import_shared_utils.REQUEST_DEPTH_HEADER}: ${requestDepth}, LWR Version: ${import_config.LWR_VERSION}`
      });
      if (!forwarded) {
        return res.status(400).send({
          error: "Bad Request",
          message: "The 'Forwarded' header is missing."
        });
      }
    }
    if (requestDepth && requestDepth > 1) {
      import_diagnostics.logger.error({
        label: "request-processor-middleware",
        message: `Lambda SSR request cycle detected: ${req.originalUrl}`
      });
      return res.status(400).send("Request depth limit reached");
    }
    handleRequestClass(req, basePath, requestClass);
    await next();
  });
}
function logRequestHeaders(headers) {
  if (import_diagnostics.logger.isDebugEnabled()) {
    for (const headerName in headers) {
      import_diagnostics.logger.debug({
        label: "request-processor-middleware",
        message: `Header ${headerName}: ${headers[headerName]}`
      });
    }
  }
}
function buildCookieMessage(req, length) {
  let msg = `Cookie length: ${length}`;
  if (length) {
    msg += `, Cookie has '__Secure-has-sid': ${!!req.cookie("__Secure-has-sid")}, Cookie has 'sid': ${!!req.cookie("sid")}`;
  }
  return msg;
}
function handleRequestClass(req, defaultBasePath, requestClass) {
  if (requestClass && typeof requestClass === "string") {
    const parsed = parseRequestClass(requestClass);
    import_diagnostics.logger.debug({
      label: "request-processor-middleware",
      message: `parsedRequestClass?.basePath: ${parsed?.basePath}`
    });
    const basePath = parsed?.basePath || defaultBasePath || "";
    req.basePath = basePath === "" || basePath.startsWith("/") ? basePath : `/${basePath}`;
    adjustExpressUrlForBasePath(req, parsed?.basePath);
  } else {
    import_diagnostics.logger.debug({
      label: "request-processor-middleware",
      message: `${import_shared_utils.MRT_REQUEST_CLASS} ignored ${requestClass ?? "no-headers"}`
    });
    req.basePath = defaultBasePath;
  }
}
function adjustExpressUrlForBasePath(req, basePath) {
  const expressRequest = req.req;
  if (!expressRequest?.url || !basePath)
    return;
  const {pathname, search} = new URL(expressRequest.url, "http://localhost");
  if (pathname.startsWith(basePath)) {
    const newPath = pathname.slice(basePath.length) || "/";
    expressRequest.url = newPath + search;
    expressRequest.originalUrl = newPath + search;
  } else {
    import_diagnostics.logger.warn({
      label: "request-processor-middleware",
      message: `The URL requested for doesn't start with the Base path`
    });
  }
}
function parseRequestClass(requestClass) {
  if (!requestClass) {
    return {};
  }
  const keyValuePairs = requestClass.split(";");
  const parsed = {};
  for (const pair of keyValuePairs) {
    const [key, value] = pair.split("=");
    parsed[key] = value;
  }
  return parsed;
}
