// @lwrjs/core/build/cjs/middleware/view-middleware.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/core/src/middleware/view-middleware.ts
__markAsModule(exports);
__export(exports, {
  viewMiddleware: () => viewMiddleware
});
var import_url = __toModule(require("url"));
var import_util = __toModule(require("util"));
var import_zlib = __toModule(require("zlib"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_instrumentation = __toModule(require("@lwrjs/instrumentation"));
var import_router = __toModule(require("@lwrjs/router"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_view_registry = __toModule(require("@lwrjs/view-registry"));
var import_compression = __toModule(require("./utils/compression.cjs"));
var import_error_handling = __toModule(require("./utils/error-handling.cjs"));
function createViewMiddleware(route, errorRoutes, context, viewHandler) {
  const errorRoute = errorRoutes.find((route2) => route2.status === 500);
  const appConfig = context.appConfig;
  const {environment: environmentConfig} = appConfig;
  const mrtCompressionThreshold = (0, import_compression.getMrtCompressionThreshold)();
  return async (req, res) => {
    if (!req.validateEnvironmentRequest(appConfig)) {
      res.status(400);
      res.send(import_diagnostics.descriptions.INVALID.INVALID_ENVIRONMENT(req.params.environment));
      return;
    }
    if (!req.validateJsonRequest()) {
      res.status(400);
      res.send(import_diagnostics.descriptions.INVALID.INVALID_JSON());
      return;
    }
    const {runtimeEnvironment, runtimeParams: defaultRuntimeParams} = req.getRuntimeContext(context.runtimeEnvironment);
    if (!defaultRuntimeParams.environment && environmentConfig?.default) {
      defaultRuntimeParams.environment = environmentConfig.default;
    }
    const viewRequest = {
      url: req.originalUrl,
      params: req.params,
      query: req.query,
      requestPath: req.path
    };
    const runtimeParams = {
      ...defaultRuntimeParams,
      url: viewRequest.url,
      params: viewRequest.params,
      query: viewRequest.query,
      cookie: req.headers?.cookie,
      trueClientIP: req.headers?.[import_shared_utils.TRUE_CLIENT_IP],
      correlationID: req.headers?.[import_shared_utils.CORRELATION_ID],
      coreProxy: req.getCoreProxy(appConfig.coreProxy ?? void 0, route.bootstrap?.proxyForSSR)
    };
    const resolve = req.isJsonRequest() ? viewHandler.getViewJson : viewHandler.getViewContent;
    let viewResponse;
    let resolvedRoute;
    let traceId;
    let headers = {};
    try {
      viewResponse = await (0, import_instrumentation.getTracer)().trace({
        name: import_instrumentation.RequestHandlerSpan.GetView,
        attributes: {
          view: route.id,
          url: req.originalUrl,
          host: runtimeParams.host,
          basePath: runtimeParams.basePath,
          locale: runtimeParams.locale
        }
      }, async (span) => {
        runtimeParams.viewSpanID = span.spanId;
        traceId = span.traceId;
        headers = (0, import_shared_utils.getTraceHeaders)(runtimeParams, span);
        const res2 = await resolve.call(viewHandler, viewRequest, route, runtimeEnvironment, runtimeParams);
        span.setAttributes({
          size: byteSize(res2.body)
        });
        return res2;
      });
      resolvedRoute = route;
    } catch (err) {
      if (!errorRoute || runtimeEnvironment.debug) {
        throw err;
      }
      import_diagnostics.logger.error(`[view-middleware] Route Error ${req.originalUrl}`);
      import_diagnostics.logger.error(err);
      viewResponse = await (0, import_instrumentation.getTracer)().trace({
        name: import_instrumentation.RequestHandlerSpan.GetView,
        attributes: {
          view: errorRoute.id
        }
      }, () => {
        return resolve.call(viewHandler, viewRequest, errorRoute, runtimeEnvironment, runtimeParams);
      });
      resolvedRoute = errorRoute;
    } finally {
      if (Object.keys(headers).length)
        res.set(headers);
      if (traceId?.length) {
        const serverTiming = (0, import_instrumentation.getTraceCollector)().getSpansInTrace(traceId);
        if (serverTiming) {
          res.set({
            "Server-Timing": serverTiming
          });
        }
        (0, import_instrumentation.getTraceCollector)().dropTrace(traceId);
      }
    }
    if (req.isSiteGeneration()) {
      res.setSiteGenerationMetadata(viewResponse.metadata);
    }
    res.type("text/html");
    if ((0, import_shared_utils.isLocalDev)() || (0, import_shared_utils.isLocalPreview)()) {
      if ((0, import_shared_utils.isLocalAuthEnabled)()) {
        res.setHeader("Set-Cookie", "LocalAuthenticationEnabled=1;");
      } else {
        res.setHeader("Set-Cookie", "LocalAuthenticationEnabled=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT;");
      }
    }
    if (viewResponse.headers) {
      if ((0, import_shared_utils.isLocalDev)()) {
        delete viewResponse.headers["content-security-policy"];
      }
      res.set(viewResponse.headers);
    }
    if (!res.hasHeader("cache-control")) {
      const cacheTtl = (0, import_shared_utils.shortestTtl)(viewResponse.cache?.ttl, resolvedRoute.cache?.ttl);
      if (cacheTtl) {
        res.setHeader("cache-control", `public, max-age=${cacheTtl}`);
      }
    }
    let status = resolvedRoute.status || viewResponse.status || 200;
    if (viewResponse.status === 301 || viewResponse.status === 302) {
      status = viewResponse.status;
    }
    res.status(status);
    const viewResponseBody = viewResponse.body;
    let shouldCompress, viewResponseSize;
    if ((0, import_shared_utils.isLambdaEnv)() && typeof viewResponseBody === "string") {
      viewResponseSize = Buffer.byteLength(viewResponseBody, "utf-8");
      shouldCompress = Buffer.byteLength(viewResponseBody, "utf-8") >= mrtCompressionThreshold;
    }
    if (shouldCompress) {
      const start = performance.now();
      let compressionMethod;
      if (req.headers["accept-encoding"]?.includes("gzip")) {
        compressionMethod = import_zlib.gzipSync;
        res.setHeader("Content-Encoding", "gzip");
      } else {
        res.setHeader("Content-Encoding", "br");
        compressionMethod = import_zlib.brotliCompressSync;
      }
      const compressedView = compressionMethod(viewResponseBody);
      const end = performance.now();
      import_diagnostics.logger.warn(`View size (${viewResponseSize}) compressed due to Lambda response limit. Compression took ${Math.round(end - start)} ms`);
      res.send(compressedView);
    } else {
      res.send(viewResponse.body);
    }
  };
}
function createConfigMiddleware(routes, context, viewHandler) {
  return async (req, res) => {
    const {runtimeEnvironment, runtimeParams: defaultRuntimeParams} = req.getRuntimeContext(context.runtimeEnvironment);
    const {appId, encodedViewPath} = req.params;
    const route = routes.find((route2) => route2.id === appId);
    if (!route) {
      res.status(404).send(import_diagnostics.descriptions.UNRESOLVABLE.APP_CONFIG(appId));
      return;
    }
    const url = (0, import_shared_utils.decodeViewPath)(encodedViewPath);
    let requestPath = url;
    const query = {};
    if (url.indexOf("?") !== -1) {
      requestPath = url.substring(0, url.indexOf("?"));
      const searchParams = new import_url.URLSearchParams(url.substring(url.indexOf("?")));
      for (const [key, value] of searchParams.entries()) {
        query[key] = value;
      }
    }
    const originalUrl = req.originalUrl;
    const routePath = route.path;
    const params = (0, import_shared_utils.extractRequestParams)(routePath, url, req.params);
    const viewRequest = {
      url,
      originalUrl,
      params,
      query,
      requestPath
    };
    const runtimeParams = {
      ...defaultRuntimeParams,
      url: viewRequest.url,
      params: viewRequest.params,
      query: viewRequest.query
    };
    const viewResponse = await viewHandler.getViewConfiguration(viewRequest, route, runtimeEnvironment, runtimeParams);
    if (!viewResponse) {
      res.status(404).send(import_diagnostics.descriptions.UNRESOLVABLE.VIEW_CONFIG(url));
      return;
    }
    if (!res.hasHeader("cache-control")) {
      const cacheTtl = (0, import_shared_utils.shortestTtl)(viewResponse.cache?.ttl, route.cache?.ttl);
      if (cacheTtl) {
        res.setHeader("cache-control", `public, max-age=${cacheTtl}`);
      }
    }
    res.type("application/javascript");
    res.status(viewResponse.status || 200);
    res.send(viewResponse.body);
  };
}
function createNotFoundMiddleware(errorRoutes, context, viewHandler) {
  const notFoundRoute = errorRoutes.find((route) => route.status === 404);
  if (notFoundRoute) {
    return createViewMiddleware(notFoundRoute, errorRoutes, context, viewHandler);
  }
  return (req, res) => {
    res.status(404).send(import_diagnostics.descriptions.UNRESOLVABLE.VIEW(req.originalUrl));
  };
}
function viewMiddleware(app, context) {
  const {appConfig, viewRegistry, moduleRegistry, routeHandlers} = context;
  const {i18n, routes, errorRoutes} = appConfig;
  const viewHandler = new import_view_registry.LwrViewHandler({viewRegistry, moduleRegistry, routeHandlers}, appConfig);
  for (const route of routes) {
    const paths = [route.path];
    const subRoutes = route.subRoutes && (0, import_router.getClientRoutes)(route.subRoutes);
    if (subRoutes) {
      const prefix = route.path === "/" ? "" : route.path;
      subRoutes.routes.forEach((subRoute) => subRoute.uri !== route.path && paths.push(`${prefix}${subRoute.uri}`));
    }
    if (i18n.uriPattern === "path-prefix") {
      const supportedStr = i18n.locales.map((l) => l.id).filter((id) => id !== i18n.defaultLocale).join("|");
      if (supportedStr) {
        const defaultLocalePaths = paths.map((routePath) => `/${i18n.defaultLocale}${routePath}`);
        paths.forEach((routePath) => {
          const localizedPath = `/:locale(${supportedStr})${routePath}`;
          paths.push(localizedPath);
        });
        addDefaultLocaleRedirects(i18n.defaultLocale, defaultLocalePaths, i18n.defaultRedirectParams, app);
      }
    }
    import_diagnostics.logger.debug({label: `view-middleware`, message: `Add view paths ${paths}`});
    app.get(paths, (0, import_error_handling.handleErrors)(createViewMiddleware(route, errorRoutes, context, viewHandler), true));
  }
  app.get((0, import_shared_utils.getClientBootstrapConfigurationRoutes)(), (0, import_error_handling.handleErrors)(createConfigMiddleware(routes, context, viewHandler)));
  app.get("/" + app.getRegexWildcard(), (0, import_error_handling.handleErrors)(createNotFoundMiddleware(errorRoutes, context, viewHandler)));
}
function addDefaultLocaleRedirects(defaultLocale, defaultLocalePaths, defaultRedirectParams, app) {
  import_diagnostics.logger.debug({label: `view-middleware`, message: `Add default localized paths ${defaultLocalePaths}`});
  app.get(defaultLocalePaths, (req, res, next) => {
    if (!req.originalUrl?.startsWith(`/${defaultLocale}`)) {
      import_diagnostics.logger.warn({
        label: "view-middleware",
        message: `Attempted to redirect to a URL that did not start with the default locale: /${defaultLocale} ${req.originalUrl}`
      });
      return next();
    }
    const [originalPath, queryString] = req.originalUrl.split("?");
    let modifiedPath = originalPath.replace(`/${defaultLocale}`, "");
    if (req.basePath) {
      modifiedPath = `${req.basePath}${modifiedPath}`;
    }
    const queryParams = new import_url.URLSearchParams(queryString);
    if (defaultRedirectParams) {
      Object.entries(defaultRedirectParams).forEach(([key, value]) => {
        if (queryParams.has(key) && queryParams.get(key) !== value) {
          queryParams.set(key, value);
        } else if (!queryParams.has(key)) {
          queryParams.set(key, value);
        }
      });
    }
    modifiedPath = queryParams.toString() ? `${modifiedPath}?${queryParams.toString()}` : modifiedPath;
    res.setHeader("Location", modifiedPath);
    return res.sendStatus(301);
  });
}
function byteSize(body) {
  if (Buffer.isBuffer(body)) {
    return body.length;
  } else if (typeof body === "string") {
    return new import_util.TextEncoder().encode(body).length;
  }
  return -1;
}
