// @lwrjs/core/build/es/middleware/request-processor-middleware.js
import { logger } from '@lwrjs/diagnostics';
import { CORRELATION_ID, MRT_REQUEST_CLASS, REQUEST_DEPTH_HEADER, TRUE_CLIENT_IP, isLambdaEnv, parseRequestDepth, } from '@lwrjs/shared-utils';
import { LWR_VERSION } from '@lwrjs/config';
export function requestProcessorMiddleware(app, context) {
    const { basePath } = context.runtimeEnvironment;
    app.use(async (req, res, next) => {
        if (!req.headers) {
            req.basePath = basePath;
            await next();
            return;
        }
        logRequestHeaders(req.headers);
        const requestClass = req.headers[MRT_REQUEST_CLASS];
        const requestDepth = parseRequestDepth(req.headers, req.query);
        if (isLambdaEnv()) {
            const forwarded = req.headers['forwarded'];
            const forwardedProto = req.headers['x-forwarded-proto'];
            const host = req.headers['host'];
            const trueClientIP = req.headers[TRUE_CLIENT_IP];
            const correlationID = req.headers[CORRELATION_ID];
            const protocol = req.protocol;
            const cookieLength = req.headers['cookie']?.length || 0;
            const cookieMsg = buildCookieMessage(req, cookieLength);
            logger.info({
                label: 'request-processor-middleware',
                message: `Original Url: ${req.originalUrl}, Forwarded: ${forwarded}, X-Forwarded-Proto: ${forwardedProto}, Host: ${host}, Protocol: ${protocol}, ${cookieMsg}, ${MRT_REQUEST_CLASS}: ${requestClass}, ${TRUE_CLIENT_IP}: ${trueClientIP}, ${CORRELATION_ID}: ${correlationID}, ${REQUEST_DEPTH_HEADER}: ${requestDepth}, LWR Version: ${LWR_VERSION}`,
            });
            if (!forwarded) {
                return res.status(400).send({
                    error: 'Bad Request',
                    message: "The 'Forwarded' header is missing.",
                });
            }
        }
        if (requestDepth && requestDepth > 1) {
            logger.error({
                label: 'request-processor-middleware',
                message: `Lambda SSR request cycle detected: ${req.originalUrl}`,
            });
            return res.status(400).send('Request depth limit reached');
        }
        handleRequestClass(req, basePath, requestClass);
        await next();
    });
}
// --- Helper functions ---
function logRequestHeaders(headers) {
    if (logger.isDebugEnabled()) {
        for (const headerName in headers) {
            logger.debug({
                label: 'request-processor-middleware',
                message: `Header ${headerName}: ${headers[headerName]}`,
            });
        }
    }
}
function buildCookieMessage(req, length) {
    let msg = `Cookie length: ${length}`;
    if (length) {
        msg += `, Cookie has '__Secure-has-sid': ${!!req.cookie('__Secure-has-sid')}, Cookie has 'sid': ${!!req.cookie('sid')}`;
    }
    return msg;
}
function handleRequestClass(req, defaultBasePath, requestClass) {
    if (requestClass && typeof requestClass === 'string') {
        const parsed = parseRequestClass(requestClass);
        logger.debug({
            label: 'request-processor-middleware',
            message: `parsedRequestClass?.basePath: ${parsed?.basePath}`,
        });
        const basePath = parsed?.basePath || defaultBasePath || '';
        req.basePath = basePath === '' || basePath.startsWith('/') ? basePath : `/${basePath}`;
        adjustExpressUrlForBasePath(req, parsed?.basePath);
    }
    else {
        logger.debug({
            label: 'request-processor-middleware',
            message: `${MRT_REQUEST_CLASS} ignored ${requestClass ?? 'no-headers'}`,
        });
        req.basePath = defaultBasePath;
    }
}
function adjustExpressUrlForBasePath(req, basePath) {
    const expressRequest = req.req;
    if (!expressRequest?.url || !basePath)
        return;
    const { pathname, search } = new URL(expressRequest.url, 'http://localhost');
    if (pathname.startsWith(basePath)) {
        const newPath = pathname.slice(basePath.length) || '/';
        expressRequest.url = newPath + search;
        expressRequest.originalUrl = newPath + search;
    }
    else {
        logger.warn({
            label: 'request-processor-middleware',
            message: `The URL requested for doesn't start with the Base path`,
        });
    }
}
/**
 * Parse the basePath passed via the X-Mobify-Request-Class header.
 * Example: basePath=/token
 *
 * basePath: The dynamic base path
 *  '' or undefined -> LWR basePath ''
 *  token or /token -> LWR basePath '/token'
 *
 */
function parseRequestClass(requestClass) {
    // If there is no requestClass do not bother parsing
    if (!requestClass) {
        return {};
    }
    // Split the header into individual key-value pairs
    const keyValuePairs = requestClass.split(';');
    // Create an object to store the parsed values
    const parsed = {};
    // Iterate through the key-value pairs and populate the parsed object
    for (const pair of keyValuePairs) {
        const [key, value] = pair.split('=');
        parsed[key] = value;
    }
    return parsed;
}
