// @lwrjs/diagnostics/build/es/descriptions/unresolvable.js
/* istanbul ignore file */
export const unresolvable = {
    ASSET: (assetPath) => `404: Could not find asset at "${escape(assetPath)}"`,
    DIR_ALIAS: (assetPath) => `404: Unable to normalize "${escape(assetPath)}", no alias replacement found`,
    MODULE: (moduleSpecifier) => `404: Could not find module with specifier "${escape(moduleSpecifier)}"`,
    MODULE_ENTRY: (moduleSpecifier) => `404: Could not find module entry with specifier "${escape(moduleSpecifier)}"`,
    BUNDLE: (specifier) => `404: Could not find bundle with specifier "${escape(specifier)}"`,
    LWC_MODULE: (name) => `404: Could not find ${escape(name)}`,
    SIGNED_MODULE: (moduleSpecifier, signature) => `404: Could not find module with specifier "${escape(moduleSpecifier)}" and signature "${escape(signature)}"`,
    RESOURCE: (resourceSpecifier, resourceVersion) => `404: Could not find resource with specifier "${escape(resourceSpecifier)}" and version "${escape(resourceVersion)}"`,
    VIEW: (viewPath) => `404: Could not find view at "${escape(viewPath)}"`,
    VIEW_CONFIG: (configPath) => `404: LWR configuration for viewPath "${escape(configPath)}" is not available`,
    APP_CONFIG: (appId) => `404: LWR configuration for appId "${escape(appId)}" is not available`,
    LABEL_MODULE: (file, error) => `Invalid JSON: Labels file ${escape(file)} must be a JSON object with string values or nested string values\n${error}`,
    ROUTES_MODULE: (file, error) => `Invalid JSON: Route config file ${escape(file)} is invalid\n${error}`,
};
