// @lwrjs/api/build/es/localdev/providers/sfdx-module-provider.js
import path from 'path';
import fs from 'fs';
import LWCModuleProvider from '@lwrjs/lwc-module-provider';
import { explodeSpecifier } from '@lwrjs/shared-utils';
import { logger } from '@lwrjs/diagnostics';
const SFDX_SCOPE = 'c/';
const DEFAULT_MODULE_DIR = path.join(process.cwd(), 'force-app', 'main', 'default', 'lwc');
function resolveResource(name) {
    if (name.indexOf('#') > -1) {
        const [component, resource] = name.split('#');
        return { component, resource };
    }
    if (name.indexOf('/') > -1) {
        const [component, ...resource] = name.split('/');
        return { component, resource: resource.join('/') };
    }
    return { component: name, resource: undefined };
}
export default class SFDXProvider extends LWCModuleProvider {
    name = 'sfdx-module-provider';
    namespacePaths;
    constructor(options, context) {
        super(options, context);
        this.namespacePaths =
            options.namespacePaths && options.namespacePaths.length > 0
                ? options.namespacePaths
                : [DEFAULT_MODULE_DIR];
    }
    async getModuleEntry({ specifier }) {
        if (!specifier.startsWith(SFDX_SCOPE)) {
            return undefined;
        }
        const { name } = explodeSpecifier(specifier);
        const { component, resource } = resolveResource(name);
        // Check to see if we can find the parent directory locally (indicating we should have the component source)
        // this is more safe as some of the entires look like: 'c/dev#dev.scoped.css?scoped=true' (which wouldn't map to a file)
        const parentDir = this.namespacePaths.find((namespacePath) => {
            if (!fs.existsSync(path.join(namespacePath, component))) {
                return false;
            }
            return true;
        });
        if (!parentDir) {
            // if we can't find local c/src files, output a warning to the user and return undefined here
            // we could also throw here but I think returning undefined is safer for now
            logger.warn({
                label: `local-dev`,
                message: `Missing source files for module '${specifier}' in the configured package directories`,
            }, {
                packageDirectories: this.namespacePaths,
            });
            return undefined;
        }
        return {
            id: specifier,
            specifier,
            entry: path.join(parentDir, component, resource ?? `${component}.js`),
            version: '',
            scope: process.cwd(),
        };
    }
}
