// @lwrjs/instrumentation/build/cjs/aws-exporter.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/instrumentation/src/aws-exporter.ts
__markAsModule(exports);
__export(exports, {
  AWSExporter: () => AWSExporter,
  convertToReadableSpan: () => convertToReadableSpan
});
var import_core = __toModule(require("@opentelemetry/core"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_metrics_sender = __toModule(require("@salesforce/pwa-kit-runtime/utils/ssr-server/metrics-sender.cjs"));
var import_spans = __toModule(require("./spans.cjs"));
var DEFAULT_DIMENSIONS = {
  Project: process.env.MOBIFY_PROPERTY_ID,
  Target: process.env.DEPLOY_TARGET
};
var SUPPORTED_SPANS = new Set([
  import_spans.RequestHandlerSpan.GetView,
  import_spans.RequestHandlerSpan.GetAsset,
  import_spans.ViewSpan.ExecuteRouteHandler,
  import_spans.ViewSpan.GetServerData
]);
function toDurationMetric(name, span) {
  return {
    name,
    dimensions: DEFAULT_DIMENSIONS,
    timestamp: new Date((0, import_core.hrTimeToTimeStamp)(span.startTime)),
    unit: "Milliseconds",
    value: (0, import_core.hrTimeToMilliseconds)(span.duration)
  };
}
function toCountMetric(name, span) {
  return {
    name,
    dimensions: DEFAULT_DIMENSIONS,
    timestamp: new Date((0, import_core.hrTimeToTimeStamp)(span.startTime)),
    unit: "Count",
    value: 1
  };
}
function getErrorType(span) {
  if (span.status.code > 0 && span.events.length === 0) {
    return;
  }
  for (const event of span.events) {
    if (event.name === "exception" && event.attributes) {
      if (typeof event.attributes["exception.type"] === "string") {
        return event.attributes["exception.type"];
      }
    }
  }
}
function isFrameworkError(errorType) {
  return errorType !== "DiagnosticsError" && errorType !== "LwrApplicationError" && errorType !== "LwrInvalidError" && errorType !== "LwrStatusError";
}
function toCustomMetrics(span) {
  const metrics = [];
  const errorType = getErrorType(span);
  if (span.name === import_spans.RequestHandlerSpan.GetView) {
    metrics.push(toDurationMetric("ResolveViewTime", span));
    if (errorType && isFrameworkError(errorType)) {
      metrics.push(toCountMetric("ResolveViewErrors", span));
    } else if (errorType) {
      metrics.push(toCountMetric("ResolveViewAppErrors", span));
    }
  }
  if (span.name === import_spans.ViewSpan.ExecuteRouteHandler) {
    metrics.push(toDurationMetric("RouteHandlerTime", span));
    if (errorType && errorType === "LwrApplicationError") {
      metrics.push(toCountMetric("RouteHandlerErrors", span));
    }
  }
  if (span.name === import_spans.RequestHandlerSpan.GetAsset) {
    metrics.push(toDurationMetric("ResolveAssetTime", span));
    if (errorType && isFrameworkError(errorType)) {
      metrics.push(toCountMetric("ResolveAssetErrors", span));
    } else if (errorType) {
      metrics.push(toCountMetric("ResolveAssetAppErrors", span));
    }
  }
  if (span.name === import_spans.ViewSpan.GetServerData) {
    metrics.push(toDurationMetric("ExternalRequestTime", span));
    if (errorType && isFrameworkError(errorType)) {
      metrics.push(toCountMetric("ExternalRequestFailed", span));
    } else if (errorType) {
      metrics.push(toCountMetric("ExternalAppRequestFailed", span));
    }
    if ((0, import_shared_utils.getFeatureFlags)().DATA_PROVIDER_METRICS) {
      const dataProviderMetrics = extractDataProviderMetrics(span);
      metrics.push(...dataProviderMetrics);
    }
  }
  return metrics;
}
function extractDataProviderMetrics(span) {
  const metrics = [];
  const attributes = span.attributes;
  const dataProviderSuccess = attributes.dataProviderSuccess;
  const dataProviderWarnings = attributes.dataProviderWarnings;
  const dataProviderFailed = attributes.dataProviderFailed;
  const timestamp = new Date((0, import_core.hrTimeToTimeStamp)(span.startTime));
  const successProviders = dataProviderSuccess?.filter((name) => name && name.length > 0);
  successProviders?.forEach((providerName) => {
    metrics.push({
      name: "DataProviderSuccess",
      dimensions: {
        ...DEFAULT_DIMENSIONS,
        DataProviderName: providerName
      },
      timestamp,
      unit: "Count",
      value: 1
    });
  });
  const warningProviders = dataProviderWarnings?.filter((name) => name && name.length > 0);
  warningProviders?.forEach((providerName) => {
    metrics.push({
      name: "DataProviderFailure",
      dimensions: {
        ...DEFAULT_DIMENSIONS,
        DataProviderName: providerName
      },
      timestamp,
      unit: "Count",
      value: 1
    });
  });
  const failedProviders = dataProviderFailed?.filter((name) => name && name.length > 0);
  failedProviders?.forEach((providerName) => {
    metrics.push({
      name: "DataProviderFailure",
      dimensions: {
        ...DEFAULT_DIMENSIONS,
        DataProviderName: providerName
      },
      timestamp,
      unit: "Count",
      value: 1
    });
  });
  return metrics;
}
function convertToReadableSpan(span) {
  if (!span) {
    return {};
  }
  return {
    traceId: span.spanContext?.().traceId,
    parentId: span.parentSpanId,
    traceState: span.spanContext?.().traceState?.serialize(),
    name: span.name,
    id: span.spanContext?.().spanId,
    kind: span.kind,
    ...span.startTime ? {
      timestamp: (0, import_core.hrTimeToTimeStamp)(span.startTime)
    } : {},
    ...span.duration ? {
      duration: (0, import_core.hrTimeToMilliseconds)(span.duration)
    } : {},
    attributes: span.attributes,
    status: span.status,
    events: span.events,
    links: span.links,
    start_time: span.startTime,
    end_time: span.endTime,
    forwardTrace: process.env.DISABLE_B3_TRACING?.toLowerCase() !== "true" && process.env.MRT_B3_TRACING?.toLowerCase() !== "false" && import_spans.BaseDocRequestSpans.has(span.name)
  };
}
var AWSExporter = class {
  constructor() {
    this.metricsSender = import_metrics_sender.MetricsSender.getSender();
    this.isShutdown = false;
    this.isLambda = process.env.AWS_LAMBDA_FUNCTION_NAME;
  }
  export(spans, done) {
    if (this.isShutdown) {
      done({
        code: import_core.ExportResultCode.FAILED,
        error: new Error("Failed to export spans: Exporter shutdown")
      });
      return;
    }
    this.log(spans);
    if (!this.isLambda) {
      done({code: import_core.ExportResultCode.SUCCESS});
      return;
    }
    this.send(spans.filter(({name}) => SUPPORTED_SPANS.has(name)), done);
  }
  log(spans) {
    for (const span of spans) {
      console.log(JSON.stringify(convertToReadableSpan(span)));
    }
  }
  send(spans, done) {
    this.metricsSender.send(spans.map(toCustomMetrics).flat());
    done({code: import_core.ExportResultCode.SUCCESS});
  }
  async shutdown() {
    this.isShutdown = true;
  }
};
