// @lwrjs/instrumentation/build/cjs/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/instrumentation/src/index.ts
__markAsModule(exports);
__export(exports, {
  BundleSpan: () => import_spans.BundleSpan,
  ConfigSpan: () => import_spans.ConfigSpan,
  CoreSpan: () => import_spans.CoreSpan,
  LambdaSpan: () => import_spans.LambdaSpan,
  RequestHandlerSpan: () => import_spans.RequestHandlerSpan,
  ViewSpan: () => import_spans.ViewSpan
});
__exportStar(exports, __toModule(require("./tracer.cjs")));
__exportStar(exports, __toModule(require("./trace-collector.cjs")));
__exportStar(exports, __toModule(require("./cache-store.cjs")));
var import_spans = __toModule(require("./spans.cjs"));
